common/scala/src/main/scala/org/apache/openwhisk/core/database/mongodb/MongoDBArtifactStore.scala

```diff
@@ -0,0 +1,661 @@
+import org.apache.openwhisk.core.entity.Attachments.Attached
+import org.apache.openwhisk.core.entity.{DocId, DocInfo, DocRevision, DocumentReader, UUID}
+import org.apache.openwhisk.http.Messages
+import org.bson.json.{JsonMode, JsonWriterSettings}
+import org.mongodb.scala.bson.BsonString
+import org.mongodb.scala.bson.collection.immutable.Document
+import org.mongodb.scala.gridfs.{GridFSBucket, GridFSFile, MongoGridFSException}
+import org.mongodb.scala.model._
+import org.mongodb.scala.{MongoClient, MongoCollection, MongoException}
+import spray.json._
+
+import scala.concurrent.Future
+import scala.util.Try
+
+object MongoDBArtifactStore {
+  val _computed = "_computed"
+}
+
+/**
+ * Basic client to put and delete artifacts in a data store.
+ *
+ * @param client the mongodb client to access database
+ * @param dbName the name of the database to operate on
+ * @param collName the name of the collection to operate on
+ * @param documentHandler helper class help to simulate the designDoc of CouchDB
+ * @param viewMapper helper class help to simulate the designDoc of CouchDB
+ */
+class MongoDBArtifactStore[DocumentAbstraction <: DocumentSerializer](client: MongoClient,
+                                                                      dbName: String,
+                                                                      collName: String,
+                                                                      documentHandler: DocumentHandler,
+                                                                      viewMapper: MongoDBViewMapper,
+                                                                      val inliningConfig: InliningConfig,
+                                                                      val attachmentStore: Option[AttachmentStore])(
+  implicit system: ActorSystem,
+  val logging: Logging,
+  jsonFormat: RootJsonFormat[DocumentAbstraction],
+  val materializer: ActorMaterializer,
+  docReader: DocumentReader)
+    extends ArtifactStore[DocumentAbstraction]
+    with DocumentProvider
+    with DefaultJsonProtocol
+    with AttachmentSupport[DocumentAbstraction] {
+
+  import MongoDBArtifactStore._
+
+  protected[core] implicit val executionContext = system.dispatcher
+
+  private val mongodbScheme = "mongodb"
+  val attachmentScheme: String = attachmentStore.map(_.scheme).getOrElse(mongodbScheme)
+
+  private val database = client.getDatabase(dbName)
+  private val collection = getCollectionAndCreateIndexes
+  private val gridFSBucket = GridFSBucket(database, collName)
+
+  private val jsonWriteSettings = JsonWriterSettings.builder().outputMode(JsonMode.RELAXED).build
+
+  // MongoDB doesn't support using `$` as the first char of field name, so below two fields needs to be encoded first
+  private val fieldsNeedEncode = Seq("annotations", "parameters")
+
+  override protected[database] def put(d: DocumentAbstraction)(implicit transid: TransactionId): Future[DocInfo] = {
```

Comment:
I'm unsure how this variable is generally named.

Looking at called methods (e.g. `failed`, `finished`), could we rename this to `transaction`?

Reply:
you mean the `transid` here? such name are also used in other places